File: app/src/main/java/dev/ragnarok/fenrir/fragment/base/PlaceSupportMvpFragment.java
package dev.ragnarok.fenrir.fragment.base;

import android.Manifest;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collections;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.activity.SendAttachmentsActivity;
import dev.ragnarok.fenrir.adapter.AttachmentsViewBinder;
import dev.ragnarok.fenrir.adapter.listener.OwnerClickListener;
import dev.ragnarok.fenrir.dialog.PostShareDialog;
import dev.ragnarok.fenrir.domain.ILikesInteractor;
import dev.ragnarok.fenrir.fragment.search.SearchContentType;
import dev.ragnarok.fenrir.fragment.search.criteria.BaseSearchCriteria;
import dev.ragnarok.fenrir.link.LinkHelper;
import dev.ragnarok.fenrir.model.Article;
import dev.ragnarok.fenrir.model.Audio;
import dev.ragnarok.fenrir.model.AudioArtist;
import dev.ragnarok.fenrir.model.AudioPlaylist;
import dev.ragnarok.fenrir.model.Commented;
import dev.ragnarok.fenrir.model.CommentedType;
import dev.ragnarok.fenrir.model.Document;
import dev.ragnarok.fenrir.model.Link;
import dev.ragnarok.fenrir.model.Market;
import dev.ragnarok.fenrir.model.MarketAlbum;
import dev.ragnarok.fenrir.model.Message;
import dev.ragnarok.fenrir.model.Peer;
import dev.ragnarok.fenrir.model.Photo;
import dev.ragnarok.fenrir.model.PhotoAlbum;
import dev.ragnarok.fenrir.model.Poll;
import dev.ragnarok.fenrir.model.Post;
import dev.ragnarok.fenrir.model.Story;
import dev.ragnarok.fenrir.model.Video;
import dev.ragnarok.fenrir.model.WallReply;
import dev.ragnarok.fenrir.model.WikiPage;
import dev.ragnarok.fenrir.mvp.core.IMvpView;
import dev.ragnarok.fenrir.mvp.presenter.base.PlaceSupportPresenter;
import dev.ragnarok.fenrir.mvp.view.IAttachmentsPlacesView;
import dev.ragnarok.fenrir.mvp.view.base.IAccountDependencyView;
import dev.ragnarok.fenrir.place.PlaceFactory;
import dev.ragnarok.fenrir.player.MusicPlaybackService;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.AppPerms;
import dev.ragnarok.fenrir.util.AssertUtils;
import dev.ragnarok.fenrir.util.CustomToast;
import dev.ragnarok.fenrir.util.Utils;

public abstract class PlaceSupportMvpFragment<P extends PlaceSupportPresenter<V>, V extends IMvpView & IAttachmentsPlacesView & IAccountDependencyView>
        extends BaseMvpFragment<P, V> implements AttachmentsViewBinder.OnAttachmentsActionCallback, IAttachmentsPlacesView, OwnerClickListener {

    private final AppPerms.doRequestPermissions requestWritePermission = AppPerms.requestPermissions(this,
            new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE},
            () -> CustomToast.CreateCustomToast(requireActivity()).showToast(R.string.permission_all_granted_text));

    @Override
    public void onOwnerClick(int ownerId) {
        callPresenter(p -> p.fireOwnerClick(ownerId));
    }

    @Override
    public void openChatWith(int accountId, int messagesOwnerId, @NonNull Peer peer) {
        PlaceFactory.getChatPlace(accountId, messagesOwnerId, peer).tryOpenWith(requireActivity());
    }

    @Override
    public void onPollOpen(@NonNull Poll apiPoll) {
        callPresenter(p -> p.firePollClick(apiPoll));
    }

    @Override
    public void onVideoPlay(@NonNull Video video) {
        callPresenter(p -> p.fireVideoClick(video));
    }

    @Override
    public void onAudioPlay(int position, @NonNull ArrayList<Audio> apiAudio) {
        callPresenter(p -> p.fireAudioPlayClick(position, apiAudio));
    }

    @Override
    public void onForwardMessagesOpen(@NonNull ArrayList<Message> messages) {
        callPresenter(p -> p.fireForwardMessagesClick(messages));
    }

    @Override
    public void onOpenOwner(int ownerId) {
        callPresenter(p -> p.fireOwnerClick(ownerId));
    }

    @Override
    public void onGoToMessagesLookup(@NonNull Message message) {
        callPresenter(p -> p.fireGoToMessagesLookup(message));
    }

    @Override
    public void goToMessagesLookupFWD(int accountId, int peerId, int messageId) {
        PlaceFactory.getMessagesLookupPlace(accountId, peerId, messageId, null).tryOpenWith(requireActivity());
    }

    @Override
    public void goWallReplyOpen(int accountId, @NonNull WallReply reply) {
        PlaceFactory.getCommentsPlace(accountId, new Commented(reply.getPostId(), reply.getOwnerId(), CommentedType.POST, null), reply.getId())
                .tryOpenWith(requireActivity());
    }

    @Override
    public void onDocPreviewOpen(@NonNull Document document) {
        callPresenter(p -> p.fireDocClick(document));
    }

    @Override
    public void onPostOpen(@NonNull Post post) {
        callPresenter(p -> p.firePostClick(post));
    }

    @Override
    public void onLinkOpen(@NonNull Link link) {
        callPresenter(p -> p.fireLinkClick(link));
    }

    @Override
    public void onUrlOpen(@NonNull String url) {
        callPresenter(p -> p.fireUrlClick(url));
    }

    @Override
    public void onFaveArticle(@NonNull Article article) {
        callPresenter(p -> p.fireFaveArticleClick(article));
    }

    @Override
    public void onShareArticle(@NonNull Article article) {
        SendAttachmentsActivity.startForSendAttachments(requireActivity(), Settings.get().accounts().getCurrent(), article);
    }

    @Override
    public void onWikiPageOpen(@NonNull WikiPage page) {
        callPresenter(p -> p.fireWikiPageClick(page));
    }

    @Override
    public void onStoryOpen(@NonNull Story story) {
        callPresenter(p -> p.fireStoryClick(story));
    }

    @Override
    public void onUrlPhotoOpen(@NonNull String url, @NonNull String prefix, @NonNull String photo_prefix) {
        PlaceFactory.getSingleURLPhotoPlace(url, prefix, photo_prefix).tryOpenWith(requireActivity());
    }

    @Override
    public void openStory(int accountId, @NonNull Story story) {
        PlaceFactory.getHistoryVideoPreviewPlace(accountId, new ArrayList<>(Collections.singleton(story)), 0).tryOpenWith(requireActivity());
    }

    @Override
    public void onAudioPlaylistOpen(@NonNull AudioPlaylist playlist) {
        callPresenter(p -> p.fireAudioPlaylistClick(playlist));
    }

    @Override
    public void onWallReplyOpen(@NonNull WallReply reply) {
        callPresenter(p -> p.fireWallReplyOpen(reply));
    }

    @Override
    public void openAudioPlaylist(int accountId, @NonNull AudioPlaylist playlist) {
        PlaceFactory.getAudiosInAlbumPlace(accountId, playlist.getOwnerId(), playlist.getId(), playlist.getAccess_key()).tryOpenWith(requireActivity());
    }

    @Override
    public void onPhotosOpen(@NonNull ArrayList<Photo> photos, int index, boolean refresh) {
        callPresenter(p -> p.firePhotoClick(photos, index, refresh));
    }

    @Override
    public void openPhotoAlbum(int accountId, @NonNull PhotoAlbum album) {
        PlaceFactory.getVKPhotosAlbumPlace(accountId, album.getOwnerId(), album.getId(), null).tryOpenWith(requireActivity());
    }

    @Override
    public void onPhotoAlbumOpen(@NonNull PhotoAlbum album) {
        callPresenter(p -> p.firePhotoAlbumClick(album));
    }

    @Override
    public void onMarketAlbumOpen(@NonNull MarketAlbum market_album) {
        callPresenter(p -> p.fireMarketAlbumClick(market_album));
    }

    @Override
    public void onMarketOpen(@NonNull Market market) {
        callPresenter(p -> p.fireMarketClick(market));
    }

    @Override
    public void onArtistOpen(@NonNull AudioArtist artist) {
        callPresenter(p -> p.fireArtistClick(artist));
    }

    @Override
    public void openLink(int accountId, @NonNull Link link) {
        LinkHelper.openLinkInBrowser(requireActivity(), link.getUrl());
    }

    @Override
    public void openUrl(int accountId, @NonNull String url) {
        PlaceFactory.getExternalLinkPlace(accountId, url)
                .tryOpenWith(requireActivity());
    }

    @Override
    public void openWikiPage(int accountId, @NonNull WikiPage page) {
        PlaceFactory.getExternalLinkPlace(accountId, page.getViewUrl())
                .tryOpenWith(requireActivity());
    }

    @Override
    public void toMarketAlbumOpen(int accountId, @NonNull MarketAlbum market_album) {
        PlaceFactory.getMarketPlace(accountId, market_album.getOwner_id(), market_album.getId()).tryOpenWith(requireActivity());
    }

    @Override
    public void toArtistOpen(int accountId, @NonNull AudioArtist artist) {
        PlaceFactory.getArtistPlace(accountId, artist.getId(), false).tryOpenWith(requireActivity());
    }

    @Override
    public void toMarketOpen(int accountId, @NonNull Market market) {
        PlaceFactory.getMarketViewPlace(accountId, market).tryOpenWith(requireActivity());
    }

    @Override
    public void openSimplePhotoGallery(int accountId, @NonNull ArrayList<Photo> photos, int index, boolean needUpdate) {
        PlaceFactory.getSimpleGalleryPlace(accountId, photos, index, needUpdate).tryOpenWith(requireActivity());
    }

    @Override
    public void openPost(int accountId, @NonNull Post post) {
        PlaceFactory.getPostPreviewPlace(accountId, post.getVkid(), post.getOwnerId(), post).tryOpenWith(requireActivity());
    }

    @Override
    public void openDocPreview(int accountId, @NonNull Document document) {
        PlaceFactory.getDocPreviewPlace(accountId, document).tryOpenWith(requireActivity());
    }

    @Override
    public void openOwnerWall(int accountId, int ownerId) {
        PlaceFactory.getOwnerWallPlace(accountId, ownerId, null).tryOpenWith(requireActivity());
    }

    @Override
    public void openForwardMessages(int accountId, @NonNull ArrayList<Message> messages) {
        PlaceFactory.getForwardMessagesPlace(accountId, messages).tryOpenWith(requireActivity());
    }

    @Override
    public void playAudioList(int accountId, int position, @NonNull ArrayList<Audio> apiAudio) {
        MusicPlaybackService.startForPlayList(requireActivity(), apiAudio, position, false);
        if (!Settings.get().other().isShow_mini_player())
            PlaceFactory.getPlayerPlace(Settings.get().accounts().getCurrent()).tryOpenWith(requireActivity());
    }

    @Override
    public void openVideo(int accountId, @NonNull Video apiVideo) {
        PlaceFactory.getVideoPreviewPlace(accountId, apiVideo).tryOpenWith(requireActivity());
    }

    @Override
    public void openHistoryVideo(int accountId, @NonNull ArrayList<Story> stories, int index) {
        PlaceFactory.getHistoryVideoPreviewPlace(accountId, stories, index).tryOpenWith(requireActivity());
    }

    @Override
    public void openPoll(int accountId, @NonNull Poll poll) {
        PlaceFactory.getPollPlace(accountId, poll)
                .tryOpenWith(requireActivity());
    }

    @Override
    public void openComments(int accountId, Commented commented, Integer focusToCommentId) {
        PlaceFactory.getCommentsPlace(accountId, commented, focusToCommentId)
                .tryOpenWith(requireActivity());
    }

    @Override
    public void openSearch(int accountId, @SearchContentType int type, @Nullable BaseSearchCriteria criteria) {
        PlaceFactory.getSingleTabSearchPlace(accountId, type, criteria).tryOpenWith(requireActivity());
    }

    @Override
    public void goToLikes(int accountId, String type, int ownerId, int id) {
        PlaceFactory.getLikesCopiesPlace(accountId, type, ownerId, id, ILikesInteractor.FILTER_LIKES)
                .tryOpenWith(requireActivity());
    }

    @Override
    public void goToReposts(int accountId, String type, int ownerId, int id) {
        PlaceFactory.getLikesCopiesPlace(accountId, type, ownerId, id, ILikesInteractor.FILTER_COPIES)
                .tryOpenWith(requireActivity());
    }

    @Override
    public void repostPost(int accountId, @NonNull Post post) {
        PostShareDialog dialog = PostShareDialog.newInstance(accountId, post);
        getParentFragmentManager().setFragmentResultListener(PostShareDialog.REQUEST_POST_SHARE, dialog, (requestKey, result) -> {
            int method = PostShareDialog.extractMethod(result);
            int accountId1 = PostShareDialog.extractAccountId(result);
            Post post1 = PostShareDialog.extractPost(result);

            AssertUtils.requireNonNull(post1);

            switch (method) {
                case PostShareDialog.Methods.SHARE_LINK:
                    Utils.shareLink(requireActivity(), post1.generateVkPostLink(), post1.getText());
                    break;
                case PostShareDialog.Methods.REPOST_YOURSELF:
                    PlaceFactory.getRepostPlace(accountId1, null, post1).tryOpenWith(requireActivity());
                    break;
                case PostShareDialog.Methods.SEND_MESSAGE:
                    SendAttachmentsActivity.startForSendAttachments(requireActivity(), accountId1, post1);
                    break;
                case PostShareDialog.Methods.REPOST_GROUP:
                    int ownerId = PostShareDialog.extractOwnerId(result);
                    PlaceFactory.getRepostPlace(accountId1, Math.abs(ownerId), post1).tryOpenWith(requireActivity());
                    break;
            }
        });
        dialog.show(getParentFragmentManager(), "post-sharing");
    }

    @Override
    public void onRequestWritePermissions() {
        requestWritePermission.launch();
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/base/PlaceSupportMvpFragment.kt
package dev.ragnarok.fenrir.fragment.base

import android.Manifest
import android.os.Bundle
import android.view.View
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.activity.SendAttachmentsActivity
import dev.ragnarok.fenrir.dialog.PostShareDialog
import dev.ragnarok.fenrir.domain.ILikesInteractor
import dev.ragnarok.fenrir.fragment.base.AttachmentsViewBinder.OnAttachmentsActionCallback
import dev.ragnarok.fenrir.fragment.base.core.IMvpView
import dev.ragnarok.fenrir.fragment.base.listener.OwnerClickListener
import dev.ragnarok.fenrir.fragment.search.SearchContentType
import dev.ragnarok.fenrir.fragment.search.criteria.BaseSearchCriteria
import dev.ragnarok.fenrir.link.LinkHelper
import dev.ragnarok.fenrir.media.music.MusicPlaybackService.Companion.startForPlayList
import dev.ragnarok.fenrir.model.*
import dev.ragnarok.fenrir.place.PlaceFactory
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.AppPerms.requestPermissionsAbs
import dev.ragnarok.fenrir.util.Utils
import dev.ragnarok.fenrir.util.toast.CustomToast.Companion.createCustomToast
import kotlin.math.abs

abstract class PlaceSupportMvpFragment<P : PlaceSupportPresenter<V>, V> : BaseMvpFragment<P, V>(),
    OnAttachmentsActionCallback, IAttachmentsPlacesView,
    OwnerClickListener where V : IMvpView, V : IAttachmentsPlacesView {
    private val requestWritePermission = requestPermissionsAbs(
        arrayOf(
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE
        )
    ) {
        createCustomToast(requireActivity()).showToast(R.string.permission_all_granted_text)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        parentFragmentManager.setFragmentResultListener(
            PostShareDialog.REQUEST_POST_SHARE,
            this
        ) { _: String?, result: Bundle? ->
            val method = PostShareDialog.extractMethod(
                result ?: return@setFragmentResultListener
            )
            val accountId1 = PostShareDialog.extractAccountId(result)
            val post1 = PostShareDialog.extractPost(result)
            when (method) {
                PostShareDialog.Methods.SHARE_LINK -> Utils.shareLink(
                    requireActivity(),
                    post1?.generateVkPostLink(),
                    post1?.text
                )
                PostShareDialog.Methods.REPOST_YOURSELF -> PlaceFactory.getRepostPlace(
                    accountId1,
                    null,
                    post1
                ).tryOpenWith(requireActivity())
                PostShareDialog.Methods.SEND_MESSAGE -> post1?.let {
                    SendAttachmentsActivity.startForSendAttachments(
                        requireActivity(),
                        accountId1,
                        it
                    )
                }
                PostShareDialog.Methods.REPOST_GROUP -> {
                    val ownerId = PostShareDialog.extractOwnerId(result)
                    PlaceFactory.getRepostPlace(accountId1, abs(ownerId), post1)
                        .tryOpenWith(requireActivity())
                }
            }
        }
    }

    override fun onOwnerClick(ownerId: Long) {
        presenter?.fireOwnerClick(ownerId)
    }

    override fun openChatWith(accountId: Long, messagesOwnerId: Long, peer: Peer) {
        PlaceFactory.getChatPlace(accountId, messagesOwnerId, peer).tryOpenWith(requireActivity())
    }

    override fun onPollOpen(poll: Poll) {
        presenter?.firePollClick(poll)
    }

    override fun onVideoPlay(video: Video) {
        presenter?.fireVideoClick(video)
    }

    override fun onAudioPlay(position: Int, audios: ArrayList<Audio>) {
        presenter?.fireAudioPlayClick(position, audios)
    }

    override fun onForwardMessagesOpen(messages: ArrayList<Message>) {
        presenter?.fireForwardMessagesClick(messages)
    }

    override fun onOpenOwner(ownerId: Long) {
        presenter?.fireOwnerClick(ownerId)
    }

    override fun onGoToMessagesLookup(message: Message) {
        presenter?.fireGoToMessagesLookup(message)
    }

    override fun goToMessagesLookupFWD(accountId: Long, peerId: Long, messageId: Int) {
        PlaceFactory.getMessagesLookupPlace(accountId, peerId, messageId, null)
            .tryOpenWith(requireActivity())
    }

    override fun goWallReplyOpen(accountId: Long, reply: WallReply) {
        PlaceFactory.getCommentsPlace(
            accountId,
            Commented(reply.postId, reply.ownerId, CommentedType.POST, null),
            reply.getObjectId()
        )
            .tryOpenWith(requireActivity())
    }

    override fun onDocPreviewOpen(document: Document) {
        presenter?.fireDocClick(document)
    }

    override fun onPostOpen(post: Post) {
        presenter?.firePostClick(post)
    }

    override fun onLinkOpen(link: Link) {
        presenter?.fireLinkClick(link)
    }

    override fun onUrlOpen(url: String) {
        presenter?.fireUrlClick(url)
    }

    override fun onFaveArticle(article: Article) {
        presenter?.fireFaveArticleClick(article)
    }

    override fun onShareArticle(article: Article) {
        SendAttachmentsActivity.startForSendAttachments(
            requireActivity(),
            Settings.get().accounts().current,
            article
        )
    }

    override fun onWikiPageOpen(page: WikiPage) {
        presenter?.fireWikiPageClick(page)
    }

    override fun onStoryOpen(story: Story) {
        presenter?.fireStoryClick(story)
    }

    override fun onUrlPhotoOpen(url: String, prefix: String, photo_prefix: String) {
        PlaceFactory.getSingleURLPhotoPlace(url, prefix, photo_prefix)
            .tryOpenWith(requireActivity())
    }

    override fun openStory(accountId: Long, story: Story) {
        PlaceFactory.getHistoryVideoPreviewPlace(accountId, ArrayList(setOf(story)), 0)
            .tryOpenWith(requireActivity())
    }

    override fun onAudioPlaylistOpen(playlist: AudioPlaylist) {
        presenter?.fireAudioPlaylistClick(playlist)
    }

    override fun onWallReplyOpen(reply: WallReply) {
        presenter?.fireWallReplyOpen(reply)
    }

    override fun openAudioPlaylist(accountId: Long, playlist: AudioPlaylist) {
        PlaceFactory.getAudiosInAlbumPlace(
            accountId,
            playlist.getOwnerId(),
            playlist.getId(),
            playlist.getAccess_key()
        ).tryOpenWith(requireActivity())
    }

    override fun onPhotosOpen(photos: ArrayList<Photo>, index: Int, refresh: Boolean) {
        presenter?.firePhotoClick(photos, index, refresh)
    }

    override fun openPhotoAlbum(accountId: Long, album: PhotoAlbum) {
        PlaceFactory.getVKPhotosAlbumPlace(accountId, album.ownerId, album.getObjectId(), null)
            .tryOpenWith(requireActivity())
    }

    override fun onPhotoAlbumOpen(album: PhotoAlbum) {
        presenter?.firePhotoAlbumClick(album)
    }

    override fun onMarketAlbumOpen(market_album: MarketAlbum) {
        presenter?.fireMarketAlbumClick(market_album)
    }

    override fun onMarketOpen(market: Market) {
        presenter?.fireMarketClick(market)
    }

    override fun onArtistOpen(artist: AudioArtist) {
        presenter?.fireArtistClick(artist)
    }

    override fun openLink(accountId: Long, link: Link) {
        LinkHelper.openLinkInBrowser(requireActivity(), link.url)
    }

    override fun openUrl(accountId: Long, url: String) {
        PlaceFactory.getExternalLinkPlace(accountId, url)
            .tryOpenWith(requireActivity())
    }

    override fun openWikiPage(accountId: Long, page: WikiPage) {
        page.viewUrl?.let {
            PlaceFactory.getExternalLinkPlace(accountId, it)
                .tryOpenWith(requireActivity())
        }
    }

    override fun toMarketAlbumOpen(accountId: Long, market_album: MarketAlbum) {
        PlaceFactory.getMarketPlace(
            accountId,
            market_album.getOwner_id(),
            market_album.getId(),
            false
        )
            .tryOpenWith(requireActivity())
    }

    override fun toArtistOpen(accountId: Long, artist: AudioArtist) {
        PlaceFactory.getArtistPlace(accountId, artist.getId()).tryOpenWith(requireActivity())
    }

    override fun toMarketOpen(accountId: Long, market: Market) {
        PlaceFactory.getMarketViewPlace(accountId, market).tryOpenWith(requireActivity())
    }

    override fun openSimplePhotoGallery(
        accountId: Long,
        photos: ArrayList<Photo>,
        index: Int,
        needUpdate: Boolean
    ) {
        PlaceFactory.getSimpleGalleryPlace(accountId, photos, index, needUpdate)
            .tryOpenWith(requireActivity())
    }

    override fun openPost(accountId: Long, post: Post) {
        PlaceFactory.getPostPreviewPlace(accountId, post.vkid, post.ownerId, post)
            .tryOpenWith(requireActivity())
    }

    override fun openDocPreview(accountId: Long, document: Document) {
        PlaceFactory.getDocPreviewPlace(accountId, document).tryOpenWith(requireActivity())
    }

    override fun openOwnerWall(accountId: Long, ownerId: Long) {
        PlaceFactory.getOwnerWallPlace(accountId, ownerId, null).tryOpenWith(requireActivity())
    }

    override fun openForwardMessages(accountId: Long, messages: ArrayList<Message>) {
        PlaceFactory.getForwardMessagesPlace(accountId, messages).tryOpenWith(requireActivity())
    }

    override fun playAudioList(accountId: Long, position: Int, apiAudio: ArrayList<Audio>) {
        startForPlayList(requireActivity(), apiAudio, position, false)
        if (!Settings.get().other().isShow_mini_player) PlaceFactory.getPlayerPlace(
            Settings.get().accounts().current
        ).tryOpenWith(requireActivity())
    }

    override fun openVideo(accountId: Long, apiVideo: Video) {
        PlaceFactory.getVideoPreviewPlace(accountId, apiVideo).tryOpenWith(requireActivity())
    }

    override fun openHistoryVideo(accountId: Long, stories: ArrayList<Story>, index: Int) {
        PlaceFactory.getHistoryVideoPreviewPlace(accountId, stories, index)
            .tryOpenWith(requireActivity())
    }

    override fun openPoll(accountId: Long, apiPoll: Poll) {
        PlaceFactory.getPollPlace(accountId, apiPoll)
            .tryOpenWith(requireActivity())
    }

    override fun openComments(accountId: Long, commented: Commented, focusToCommentId: Int?) {
        PlaceFactory.getCommentsPlace(accountId, commented, focusToCommentId)
            .tryOpenWith(requireActivity())
    }

    override fun openSearch(
        accountId: Long,
        @SearchContentType type: Int,
        criteria: BaseSearchCriteria?
    ) {
        PlaceFactory.getSingleTabSearchPlace(accountId, type, criteria)
            .tryOpenWith(requireActivity())
    }

    override fun goToLikes(accountId: Long, type: String?, ownerId: Long, id: Int) {
        PlaceFactory.getLikesCopiesPlace(
            accountId,
            type,
            ownerId,
            id,
            ILikesInteractor.FILTER_LIKES
        )
            .tryOpenWith(requireActivity())
    }

    override fun goToReposts(accountId: Long, type: String?, ownerId: Long, id: Int) {
        PlaceFactory.getLikesCopiesPlace(
            accountId,
            type,
            ownerId,
            id,
            ILikesInteractor.FILTER_COPIES
        )
            .tryOpenWith(requireActivity())
    }

    override fun repostPost(accountId: Long, post: Post) {
        val dialog = PostShareDialog.newInstance(accountId, post)
        dialog.show(parentFragmentManager, "post-sharing")
    }

    override fun onRequestWritePermissions() {
        requestWritePermission.launch()
    }
}